### PR TITLE
Build on Scala3

### PR DESCRIPTION
**Pull request by hamnis**
* Do not use extract in Json4s as that uses Scala2 macros
* Use circe in tests in core
* Upgrade dependencies and scala2 versions
* Upgrade plugins
* Upgrade sbt
* Fix minor problems reported by Scala 3 compiler

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

**Comment by hamnis** [comment]
I signed it!

File: project/BuildSettings.scala [hunk 1]
```diff
@@ -120,10 +125,10 @@ object BuildSettings {
     mimaPreviousArtifacts := mimaPreviousVersions.map { organization.value %% name.value % _ },
     ThisBuild / mimaFailOnNoPrevious := false,
     mimaBinaryIssueFilters ++= Seq(),
-    Test / test := {
+    /*Test / test := {
```
Comment (hamnis) [inline project/BuildSettings.scala:128]: This was disabled to make sure we can run tests.
Mima reports binary compability errors on master

File: iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala [hunk 2]
```diff
@@ -12,84 +12,94 @@
...
-      (entity \ "self").extractOpt[SchemaKey].map(key => SchemaMap(key)) match {
-        case Some(map) => Right(map)
-        case None      => Left(ParseError.InvalidSchema)
-      }
-    }
-
-    def checkSchemaUri(entity: JValue): Either[ParseError, Unit] =
-      (entity \ s"$$schema").extractOpt[String] match {
-        case Some(schemaUri) if schemaUri == SelfDescribingSchema.SelfDescribingUri.toString =>
+  implicit object Json4SAttachSchemaMapComplex extends ExtractSchemaMap[Json] with ToSchema[Json] {
```
Comment (hamnis) [inline iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala:91]: needs rename

File: iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala [hunk 3]
```diff
@@ -98,25 +108,27 @@ object IgluCoreCommon {
...
       */
-    def getContent(json: JValue): JValue =
-      removeMetaFields(json)
+    def getContent(json: Json): Json =
+      json.asObject.map(_.remove("$schema").remove("self").asJson).getOrElse(json)
   }
 
   // Data
 
   implicit object Json4SAttachSchemaKeyData
```
Comment (hamnis) [inline iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala:117]: needs rename

File: iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala [hunk 4]
```diff
@@ -155,40 +167,30 @@ object IgluCoreCommon {
...
 
-  implicit object Json4SNormalizeSchema extends NormalizeSchema[JValue] {
-    def normalize(schema: SelfDescribingSchema[JValue]): JValue =
-      Extraction.decompose(schema)
-  }
-
-  implicit object Json4SNormalizeData extends NormalizeData[JValue] {
-    def normalize(instance: SelfDescribingData[JValue]): JValue =
-      Extraction.decompose(instance)
+  implicit object Json4SNormalizeSchema extends NormalizeSchema[Json] {
```
Comment (hamnis) [inline iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala:170]: needs rename

File: iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala [hunk 5]
```diff
@@ -155,40 +167,30 @@ object IgluCoreCommon {
...
+            s"$$schema" := SelfDescribingSchema.SelfDescribingUri,
+            "self" := schema.self.schemaKey
+          )
+        )
   }
 
-  object StringifySchema extends StringifySchema[JValue] {
-    def asString(container: SelfDescribingSchema[JValue]): String =
-      compact(container.normalize(Json4SNormalizeSchema))
+  implicit object Json4SNormalizeData extends NormalizeData[Json] {
```
Comment (hamnis) [inline iglu-core/src/test/scala/com/snowplowanalytics/iglu/core/IgluCoreCommon.scala:182]: needs rename

**Review by istreeter — COMMENTED** [review]
Hi @hamnis I think the reason that mima reports binary incompatibilities is because you upgraded json4s from a 3.x version to a 4.x version.  So any iglu-core method that returns a json4s class is now binary incompatible.

That's OK though, there a few different ways to proceed from here:

Option 1 is to release this a major version bump.  Then we can make any binary incompatible change we want.

Option 2 is find a way to add scala 3 support but without changing the version of json4s.

Option 3 is to simply drop support for the package iglu-core-json4s package.  This is a legacy package now anyway; historically we used json4s in a few places but now we use entirely circe.  I don't know of any open source users who depend on the json4s package, but we should check that.

Do you have any thoughts on those options?  Do you need the json4s package, or are you happy with just the circe package?

**Comment by hamnis** [comment]
I am fine with all options.

We do not need the json4s package.

**Comment by hamnis** [comment]
@istreeter would it be possible to merge this at some point?

**Comment by istreeter** [comment]
Hi @hamnis sorry for the delay.  I think this looks excellent, now that you addressed the problem with the json4s module.  I think we can even re-enable the mima test which you disabled.  I'm fairly sure everything is now backwards compatible with previous versions, so the mima test should pass.

I will merge this and make a 1.1.3 release, so you can start using it.

Thank you very much contributing this PR!  And for addressing the feedback.  This is a really nice improvement for users who are on scala 3.

**Comment by hamnis** [comment]
Excellent news. Thanks a  lot.

**Comment by istreeter** [comment]
Closed via #99

**Pull request closed without merging**